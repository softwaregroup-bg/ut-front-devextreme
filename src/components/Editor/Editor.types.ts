import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';

import type {Schema, Editors, Cards, Dropdowns, Layouts} from '../types';
export interface Props {
    object: string,
    id?: any,
    schema: Schema,
    editors?: Editors,
    type?: string,
    name?: string,
    typeField?: string,
    cards: Cards,
    layouts?: Layouts,
    layoutName?: string,
    nested?: string[],
    keyField?: string,
    resultSet?: string,
    design?: boolean,
    methods: any,
    onDropdown: (params: string[]) => Promise<Dropdowns>,
    onAdd: (params: {}) => Promise<{}>,
    onGet: (params: {}) => Promise<{}>,
    onEdit: (params: {}) => Promise<{}>
}

const styles = createStyles({
    component: {
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
