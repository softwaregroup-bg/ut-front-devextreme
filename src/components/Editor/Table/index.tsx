import React from 'react';
import {InputText, DataTable, Column, Toolbar, Button} from '../../prime';
import { v4 as uuidv4 } from 'uuid';

export const Table = React.forwardRef<{}, any>(({onChange, columns, value, dataKey = 'id'}, ref) => {
    if (typeof ref === 'function') ref(React.useState({})[0]);
    const cellEditor = React.useCallback((props, field) => <InputText
        type="text"
        autoFocus={props.index === 1}
        value={props.rowData[field]}
        onChange={({target: {value}}) => {
            const updatedValue = [...props.value];
            updatedValue[props.rowIndex][props.field] = value;
            onChange(updatedValue);
        }}
        id={`${props.rowData.id}`}
    />, [onChange]);
    const [original, setOriginal] = React.useState({index: null, value: null});

    const init = React.useCallback(({index}) => {
        setOriginal({index, value: {...value[index]}});
    }, [value, setOriginal]);

    const cancel = React.useCallback(() => {
        const restored = [...value];
        restored[original.index] = original.value;
        onChange(restored);
    }, [value, onChange]);
    const addNewRow = () => {
        const id = uuidv4();
        const newValue = Object.keys(value[0] || {}).reduce((item, key) => ({...item, [key]: '', id: id}), {});
        const updatedValue = [...value, newValue];
        onChange(updatedValue);
        setEditingRows({[id]: true});
    };
    const deleteRow = () => {
        onChange(value.filter(rowData => !selected.includes(rowData)));
    };
    const [selected, setSelected] = React.useState([]);
    const [editingRows, setEditingRows] = React.useState({});
    const onRowEditChange = (event) => {
        setEditingRows(event.data);
    };
    const leftToolbarTemplate = () => {
        return (
            <React.Fragment>
                <Button label="New" icon="pi pi-plus" className="p-button p-mr-2" onClick={addNewRow} />
                <Button label="Delete" icon="pi pi-trash" className="p-button" onClick={deleteRow} disabled={!selected || !selected.length} />
            </React.Fragment>
        );
    };
    return (
        <>
            <Toolbar className="p-mb-4" left={leftToolbarTemplate} right={null}></Toolbar>
            <DataTable
                value={value}
                editMode="row"
                dataKey={dataKey}
                className="editable-cells-table"
                onRowEditInit={init}
                onRowEditCancel={cancel}
                selection={selected}
                onSelectionChange={(e) => { setSelected(e.value); }}
                editingRows={editingRows}
                onRowEditChange={onRowEditChange}
            >
                <Column selectionMode="multiple" headerStyle={{ width: '3rem' }}></Column>
                {
                    (columns || []).map(({ field, header }) => <Column
                        key={field}
                        field={field}
                        header={header}
                        editor={props => cellEditor(props, field)}
                    />)
                }
                <Column rowEditor headerStyle={{ width: '7rem' }} bodyStyle={{ textAlign: 'center' }}></Column>
            </DataTable>
        </>
    );
});
