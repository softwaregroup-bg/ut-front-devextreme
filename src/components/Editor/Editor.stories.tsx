import React from 'react';
import { withReadme } from 'storybook-readme';

import Wrap from '../test/wrap';

// @ts-ignore: md file and not a module
import README from './README.md';
import Editor from './index';

export default {
    title: 'Editor',
    component: Editor,
    decorators: [withReadme(README)]
};

const state = {
};

export const Basic: React.FC<{}> = () => <Wrap state={state}>
    <Editor fields={[]} cards={[]} />
</Wrap>;
