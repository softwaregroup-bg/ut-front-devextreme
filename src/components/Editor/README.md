# Editor

## How to use

```jsx
import Editor from 'ut-front-devextreme/core/Editor';

<Editor className='name' />;
```

## Props

- **className** - class applied to root `Editor` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| className | string   |              | no         |
