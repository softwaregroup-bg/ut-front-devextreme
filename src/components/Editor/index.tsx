import React from 'react';
import clsx from 'clsx';
import Joi from 'joi';

import { Card, InputText, DataTable, Column, Dropdown, InputMask, InputNumber, Calendar } from '../prime';
import { Styled, StyledType } from './Editor.types';
import useForm from '../hooks/useForm';
import Controller from '../Controller';
import { joiResolver } from '@hookform/resolvers/joi';

function Currency({onChange, ref, ...props}) {
    return (
        <InputNumber
            inputRef={ref}
            onChange={e => {
                onChange?.(e.value);
            }}
            maxFractionDigits={2}
            {...props}
        />
    );
}
function Table({columns, items, ...props}) {
    const [data, setData] = React.useState([...items]);
    const dataTableFuncMap = {
        data: setData
    };

    const onEditorValueChange = (stateKey, props, value) => {
        const updatedValue = [...props.value];
        updatedValue[props.rowIndex][props.field] = value;
        dataTableFuncMap[stateKey](updatedValue);
    };
    const inputTextEditor = (stateKey, props, field) => {
        return <InputText
            type="text"
            value={props.rowData[field]}
            onChange={(event) => onEditorValueChange(stateKey, props, event.target.value)}
            id={`${props.rowData.id}`}
        />;
    };
    return (
        <DataTable value={data} editMode="cell" className="editable-cells-table">
            {(columns || []).map(({ field, header }) => <Column
                key={field}
                field={field}
                header={header}
                editor={(props) => inputTextEditor('data', props, field)}
            ></Column>
            )}
        </DataTable>
    );
}
function element(field, {type = 'string', ...props} = {}) {
    const Element: React.ElementType = {
        dropdown: Dropdown,
        mask: InputMask,
        date: Calendar,
        currency: Currency,
        table: Table
    }[type] || InputText;
    return <Element {...field} {...props}/>;
}

const Editor: StyledType = ({ classes, className, fields, cards, onSubmit, trigger, ...rest }) => {
    const schema = fields.reduce(
        (schema, {name, title, validation = Joi.string().allow('')}) => schema.append({[name]: validation.label(title)}),
        Joi.object()
    );
    const {handleSubmit, control, formState: {errors}} = useForm({resolver: joiResolver(schema)});
    const getFormErrorMessage = (name) => {
        return errors[name] && <small className="p-error">{errors[name].message}</small>;
    };
    React.useEffect(() => {
        if (trigger) trigger.current = handleSubmit(onSubmit);
    }, [trigger, handleSubmit, onSubmit]);
    return (
        <div {...rest}>
            <form onSubmit={handleSubmit(onSubmit)} className="p-grid p-m-2">
                {(cards || []).map(({id, title, className}) =>
                    <div key={id} className={clsx('p-col-12', className || 'p-xl-6')}>
                        <Card title={title} key={id} className='p-fluid'>
                            {fields.filter(({card}) => id === card).map(({name, title, editor}) =>
                                <div className="p-field p-grid" key={name}>
                                    <label className="p-col-12 p-md-4">
                                        {title}
                                    </label>
                                    <div className="p-col-12 p-md-8">
                                        <Controller
                                            control={control}
                                            name={name}
                                            render={({field}) => element({className: clsx({ 'p-invalid': errors[name] }), ...field}, editor)}
                                        />
                                    </div>
                                    {getFormErrorMessage(name)}
                                </div>
                            )}
                        </Card>
                    </div>
                )}
            </form>
        </div>
    );
};

export default Styled(Editor);
