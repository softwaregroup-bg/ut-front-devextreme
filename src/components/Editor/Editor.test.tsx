import React from 'react';

import { render } from '../test';
import Editor from './index';

describe('<Editor />', () => {
    it('render equals snapshot', () => {
        const { getByTestId } = render(<Editor fields={[]} cards={[]} onSubmit={data => {}}/>);
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
