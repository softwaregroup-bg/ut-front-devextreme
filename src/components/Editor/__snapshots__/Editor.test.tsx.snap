// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div>
    <form
      class="p-grid p-m-2"
    />
  </div>
</div>
`;
